### PR TITLE
Return JSON True instead of empty HTML for 2 requests

### PR DESCRIPTION
**Pull request by hydrargyrum**
/load_extra_conf_file and /ignore_extra_conf_file returned an empty HTML body
in case of success. Instead, return JSON body containing "true".

[Review on Reviewable](https://reviewable.io/reviews/valloric/ycmd/243)

### CODE REVIEW

**Comment by vheon** [comment]
The code looks good. Is this just for consistency with the API (let's return always a JSON) or is this driven by some other issue? 

---

Reviewed 1 of 1 files at r1.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/243)

**Comment by hydrargyrum** [comment]
Purely for consistency. I'm trying to make 401 errors (wrong HMAC signature) return JSON too.

**Comment by Valloric** [comment]
It would be nice to get some tests for this. If it's part of the API, we should make sure it doesn't regress.

---

Review status: 0 of 1 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/243)

**Comment by hydrargyrum** [comment]
I added tests for the 2 requests of the first commit.

For the HMAC handling, it seems a little harder as the HMAC verification is done in a "plugin" that doesn't seem enabled when testing. I lack knowledge if we can enable them and how to fix all tests then.

**Comment by micbou** [comment]
Send another PR for the unauthorized request.

---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

<sup>**[ycmd/tests/misc_handlers_test.py, line 58 \[r3\]](https://reviewable.io:443/reviews/valloric/ycmd/243#-K-9qZyzAxvW8WYsm8dt)** ([raw file](https://github.com/valloric/ycmd/blob/a257f2620c9f265d1de773de3bdcc87fde655fa5/ycmd/tests/misc_handlers_test.py#L58)):</sup>
Add a blank line or remove the one above.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/243)

**Comment by vheon** [comment]
Reviewed 1 of 1 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

<sup>**[ycmd/tests/misc_handlers_test.py, line 53 \[r4\]](https://reviewable.io:443/reviews/valloric/ycmd/243#-K-AAhB0xnfAAuASVrrT)** ([raw file](https://github.com/valloric/ycmd/blob/f221ad8711672e66308c8271c323a71b7d1d58b8/ycmd/tests/misc_handlers_test.py#L53)):</sup>
I know they work like they are, but to me it feels strange not having and assert there; something like:

```python
assert_that( app.post_json( '/load_extra_conf_file', data ).json, is_valid_json() )
```

But I assume that those are personal tastes. :)

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/243)

**Comment by Valloric** [comment]
Reviewed 1 of 1 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

<sup>**[ycmd/tests/misc_handlers_test.py, line 53 \[r4\]](https://reviewable.io:443/reviews/valloric/ycmd/243#-K-AAhB0xnfAAuASVrrT)** ([raw file](https://github.com/valloric/ycmd/blob/f221ad8711672e66308c8271c323a71b7d1d58b8/ycmd/tests/misc_handlers_test.py#L53)):</sup>
I agree with you; lack of an assert-like-something makes it hard to see what's actually being asserted.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/243)

**Comment by micbou** [comment]
Reviewed 1 of 1 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

<sup>**[ycmd/tests/misc_handlers_test.py, line 53 \[r4\]](https://reviewable.io:443/reviews/valloric/ycmd/243#-K-AAhB0xnfAAuASVrrT)** ([raw file](https://github.com/valloric/ycmd/blob/f221ad8711672e66308c8271c323a71b7d1d58b8/ycmd/tests/misc_handlers_test.py#L53)):</sup>
Since the four tests of this file are similars, they should be written the same way. I suggest to use the same assertion as the first one: `ok_()`.  It will remove the need to import the `hamcrest` module. The second test should also be updated.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/243)

**Comment by micbou** [comment]
Review status: 1 of 2 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

<sup>**[ycmd/tests/misc_handlers_test.py, line 45 \[r6\]](https://reviewable.io:443/reviews/valloric/ycmd/243#-K-VRTJlxwd2l4uXSbtv)** ([raw file](https://github.com/valloric/ycmd/blob/3b9d6cd2b4669d2740490c4dadaddfc4a8bd9f78/ycmd/tests/misc_handlers_test.py#L45)):</sup>
It seems that using `ok_` was a bad idea because the response is an empty dictionary. We should actually check if the object contains the `json` property:
```python
assert_that( app.post_json( '/event_notification', request_data ), has_property( 'json' ) )
```
where `assert_that` and `has_property`come from the `hamcrest` module.
Could you update (again) the four tests? It would be greatly appreciated.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/243)

**Comment by vheon** [comment]
Reviewed 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/243)

**Comment by vheon** [comment]
LGTM

**Comment by Valloric** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

<sup>**[ycmd/tests/misc_handlers_test.py, line 48 \[r7\]](https://reviewable.io:443/reviews/valloric/ycmd/243#-K-a9ngyjE8Mgb_JPzM7)** ([raw file](https://github.com/valloric/ycmd/blob/eef39174197b81a410c012961d85cff4f8211efd/ycmd/tests/misc_handlers_test.py#L48)):</sup>
I'm not sure this does the right thing. It's verifying that the WebTest response object has the "json" property, which is not what we want to test for. We want to test that the _value_ of that json is True.

In other words,  you can probably use `ok_( app.post_json( '/event_notification', event_data ).json )`.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/243)

**Comment by hydrargyrum** [comment]
handlers.py's EventNotification sometimes return an empty dict, which would not match ok_ and cause the test to fail

**Comment by vheon** [comment]
What do you mean by "sometimes"?


---

Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/243)

**Comment by hydrargyrum** [comment]
I mean this: https://github.com/Valloric/ycmd/blob/master/ycmd/handlers.py#L75

**Comment by Valloric** [comment]
> handlers.py's EventNotification sometimes return an empty dict, which would not match ok_ and cause the test to fail

Fair point, but I still don't see how checking for the `json` property on the WebTest object resolves that. We need to verify that we can parse the data coming from the handler as json.


---

Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/243)

**Comment by hydrargyrum** [comment]
It seems accessing the json property already checks if it's valid: https://github.com/Pylons/webtest/blob/master/webtest/response.py#L491

**Comment by Valloric** [comment]
Interesting, but that's an implementation detail that's not immediately obvious to the reader of the test code. As such, we shouldn't really depend on it directly. Maybe write a small wrapper function that does this, like `EnsureValidJson` and that then has a comment explaining this bit of impl detail? That could work.


---

Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/243)
